native/dgeqp3.go

```diff
@@ -0,0 +1,173 @@
+	const (
+		inb    = 1
+		inbmin = 2
+		ixover = 3
+	)
+	checkMatrix(m, n, a, lda)
+
+	if len(jpvt) != n {
+		panic(badIpiv)
+	}
+	for _, v := range jpvt {
+		if v < -1 || n <= v {
+			panic("lapack: jpvt element out of range")
+		}
+	}
+	minmn := min(m, n)
+	if len(tau) != minmn {
+		panic(badTau)
+	}
+	if len(work) < max(1, lwork) {
+		panic(badWork)
+	}
+
+	var iws, lwkopt, nb int
+	if minmn == 0 {
+		iws = 1
+		lwkopt = 1
+	} else {
+		iws = 3*n + 1
+		nb = impl.Ilaenv(inb, "DGEQRF", " ", m, n, -1, -1)
+		lwkopt = 2*n + (n+1)*nb
+	}
+	work[0] = float64(lwkopt)
+
+	if lwork == -1 {
+		return
+	}
+
+	bi := blas64.Implementation()
+
+	// Move initial columns up front.
+	var nfxd int
+	for j := 0; j < n; j++ {
+		if jpvt[j] == -1 {
+			jpvt[j] = j
+			continue
+		}
+		if j != nfxd {
+			bi.Dswap(m, a[j:], lda, a[nfxd:], lda)
+			jpvt[j], jpvt[nfxd] = jpvt[nfxd], j
+		} else {
+			jpvt[j] = j
+		}
+		nfxd++
+	}
+
+	// Factorize nfxd columns.
+	//
+	// Compute the QR factorization of nfxd columns and update remaining columns.
+	if nfxd >= 0 {
+		na := min(m, nfxd)
+		impl.Dgeqrf(m, na, a, lda, tau, work, lwork)
+		iws = max(iws, int(work[0]))
+		if na < n {
+			impl.Dormqr(blas.Left, blas.Trans, m, n-na, na, a, lda, tau[:na], a[na:], lda,
+				work, lwork)
+			iws = max(iws, int(work[0]))
+		}
+	}
+
+	if nfxd >= minmn {
+		work[0] = float64(iws)
+		return
+	}
+
+	// Factorize free columns.
+	sm := m - nfxd
+	sn := n - nfxd
+	sminmn := minmn - nfxd
+
+	// Determine the block size.
+	nb = impl.Ilaenv(inb, "DGEQRF", " ", sm, sn, -1, -1)
+	nbmin := 2
+	nx := 0
+
+	if 1 < nb && nb < sminmn {
+		// Determine when to cross over from blocked to unblocked code.
+		nx = max(0, impl.Ilaenv(ixover, "DGEQRF", " ", sm, sn, -1, -1))
+
+		if nx < sminmn {
+			// Determine if workspace is large enough for blocked code.
+			minws := 2*sn + (sn+1)*nb
+			iws = max(iws, minws)
+			if lwork < minws {
+				// Not enough workspace to use optimal nb. Reduce
+				// nb and determine the minimum value of nb.
+				nb = (lwork - 2*sn) / (sn + 1)
+				nbmin = max(2, impl.Ilaenv(inbmin, "DGEQRF", " ", sm, sn, -1, -1))
+			}
+		}
+	}
+
+	// Initialize partial column norms.
+	// The first n elements of work store the exact column norms.
+	for j := nfxd; j < n; j++ {
+		work[j] = bi.Dnrm2(sm, a[nfxd*lda+j:], lda)
+		work[n+j] = work[j]
+	}
+	j := nfxd
+	if nbmin <= nb && nb < sminmn && nx < sminmn {
+		// Use blocked code initially.
+
+		// Compute factorization.
+		var fjb int
+		for topbmn := minmn - nx; j < topbmn; j += fjb {
+			jb := min(nb, topbmn-j)
+
+			// Factorize jb columns among columns j:n.
+			fjb = impl.Dlaqps(m, n-j, j, jb, a[j:], lda, jpvt[j:], tau[j:],
+				work[j:j+n], work[j+n:j+2*n], work[2*n:2*n+jb], work[2*n+jb:], jb)
+		}
+	}
+
+	// Use unblocked code to factor the last or only block.
+	if j < minmn {
+		impl.Dlaqp2(m, n-j, j, a[j:], lda, jpvt[j:], tau[j:],
+			work[j:j+n], work[j+n:j+2*n], work[2*n:])
```

Comment:
I think you are right, though also that it doesn't matter. The Fortran doesn't define the lengths (obviously) and sets the starts where they are here. But as you say they over provide so the bounds won't matter. I've fixed them though so that if things do go out of range we don't overwrite data. (The first upper bound I think is wrong in the line above thought, given the second parameter's lower bound).